src/metabase/query_processor/middleware/optimize_temporal_filters.clj

```diff
@@ -3,65 +3,104 @@
-  the filter clause they're in?"
-  [field temporal-value]
-  (mbql.u/match-one temporal-value
-    [:relative-datetime (_ :guard #{0 :current})]
-    true
-
-    [(_ :guard #{:absolute-datetime :relative-datetime}) _ (unit :guard optimizable-units)]
-    (= (temporal-unit field) unit)))
-
-(defmethod can-optimize-filter? :default
-  [filter-clause]
-  (mbql.u/match-one filter-clause
-    [_
-     (field :guard optimizable-field?)
-     (temporal-value :guard optimizable-temporal-value?)]
-    (field-and-temporal-value-have-compatible-units? field temporal-value)))
-
-(defmethod can-optimize-filter? :between
-  [filter-clause]
-  (mbql.u/match-one filter-clause
-    [_
-     (field :guard optimizable-field?)
-     (temporal-value-1 :guard optimizable-temporal-value?)
-     (temporal-value-2 :guard optimizable-temporal-value?)]
-    (and (field-and-temporal-value-have-compatible-units? field temporal-value-1)
-         (field-and-temporal-value-have-compatible-units? field temporal-value-2))))
+(defmulti ^:private temporal-unit
+  {:arglists '([expression])}
+  mbql.u/dispatch-by-clause-name-or-class
+  :hierarchy lib.hierarchy/hierarchy)
+
+(defmethod temporal-unit :default
+  [_expr]
+  nil)
+
+(defmethod temporal-unit :field
+  [[_field _id-or-name opts]]
+  (:temporal-unit opts))
+
+(defmethod temporal-unit :relative-datetime
+  [[_relative-datetime _n unit]]
+  unit)
+
+(defmethod temporal-unit :absolute-datetime
+  [[_absolute-datetime _t unit]]
+  unit)
+
+(doseq [tag [:+ :-]]
+  (lib.hierarchy/derive tag ::arithmetic-expr))
+
+(defmethod temporal-unit ::arithmetic-expr
+  [[_tag & args]]
+  (some temporal-unit args))
+
+(defmethod temporal-unit :+
+  [[_+ & args]]
+  (some temporal-unit args))
+
+(defmulti ^:private optimizable?
+  {:arglists '([expression])}
+  mbql.u/dispatch-by-clause-name-or-class
+  :hierarchy lib.hierarchy/hierarchy)
+
+(defmethod optimizable? :default
+  [_expr]
+  false)
+
+(defmethod optimizable? :field
+  [[_field _id-or-name opts]]
+  (contains? optimizable-units (:temporal-unit opts)))
+
+(defmethod optimizable? :expression
+  [[_expression _name opts]]
+  (let [expression-type ((some-fn :effective-type :base-type) opts)]
+    (isa? expression-type :type/Temporal)))
+
+(defmethod optimizable? :relative-datetime
+  [[_relative-datetime n unit]]
+  (if unit
+    (contains? optimizable-units unit)
+    (#{0 :current} n)))
+
+(defmethod optimizable? :absolute-datetime
+  [[_absolute-datetime _t unit]]
+  (contains? optimizable-units unit))
+
+(defmethod optimizable? ::arithmetic-expr
+  [[_tag & args]]
+  (some optimizable? args))
+
+(doseq [tag [:= :!= :< :<= :> :>=]]
+  (lib.hierarchy/derive tag ::binary-filter))
+
+(defn- compatible-units? [lhs-unit rhs-unit]
+  (case [(boolean lhs-unit) (boolean rhs-unit)]
+    [true true]   (= lhs-unit rhs-unit)
+    [true false]  true
+    [false true]  true
+    [false false] false))
```

Comment:
Maybe this is just my stylistic tastes, but I'm not a fan of these sorts of "truth tables".
I'd rather see this as

```suggestion
(if (and lhs-unit rhs-unit)
  (= lhs-unit rhs-unit)
  (boolean (or lhs-unit rhs-unit)))
```

Reply:
yeah good point I'll change it.